tests/tests-registry.php

```diff
@@ -0,0 +1,268 @@
+<?php
+/**
+ * Tests for the Registry API.
+ *
+ * @group edd_registry
+ * @group edd_utils
+ */
+class EDD_Registry_Tests extends EDD_UnitTestCase {
+
+	/**
+	 * Mock registry test fixture.
+	 *
+	 * @access protected
+	 * @var    \EDD_Registry
+	 */
+	protected $mockRegistry;
+
+	/**
+	 * Set up fixtures once.
+	 */
+	public function setUp() {
+		parent::setUp();
+
+		$this->mockRegistry = $this->getMockForAbstractClass( 'EDD_Registry' );
+	}
+
+	/**
+	 * Runs after each test to reset the items array.
+	 *
+	 * @access public
+	 */
+	public function tearDown() {
+		$this->mockRegistry->_reset_items();
+
+		parent::tearDown();
+	}
+
+	/**
+	 * @covers \EDD_Registry::init()
+	 */
+	public function test_class_should_have_an_init_method() {
+		$this->assertTrue( method_exists( $this->mockRegistry, 'init' ) );
+	}
+
+	/**
+	 * @covers \EDD_Registry::add_item()
+	 * @expectedException EDD_Exception
+	 */
+	public function test_add_item_with_empty_attributes_should_return_false() {
+		$this->assertFalse( $this->mockRegistry->add_item( 'foo', array() ) );
+	}
+
```

Comment:
@DrewAPicture can you explain this test? I'm curious why we test a `false` return here? `add_item()` returns an exception if passed empty attributes, and this test passes due to the `@expectedException` tag, which is tested in `test_add_item_with_empty_attributes_should_throw_exception()` below.

Reply:
The exception doesn't get _returned_ it gets thrown. One of the primary benefits of throwing exceptions vs returning `WP_Error` objects is that it allows you to return consistent types while still flagging errors. In this case, `add_item()` consistently returns `true` or `false`.

Reply:
I understand `return` vs `throw`, but what I'm not getting is why we test for a `false` return when passing empty attributes? An exception is thrown when attributes is empty, it doesn't return false. Without the `@expectedException ` tag, the assertion fails. But if we're testing for an exception by using the phpdoc tag, how is it different than the test below that tests the same exception with `$this->setExpectedException()`?

I guess I'm asking, what's the real difference between these two tests?

Reply:
I understand your confusion.

It's about _literally_ testing the expected _default_ in lieu of any other known or unknown factors. In the case of `add_item()` the default is `false` and we're testing that the default is false. Code would never be literally checking for a `false` value against `add_item()` because of the need for a try/catch, it's merely testing that the code works as written in the spirit of completeness. We don't always test code 100 percent for real world use, some of it is just to test that the code is doing what it's supposed to be doing.

Reply:
Ahh, I see. Thanks for explaining!